TombEngine/Renderer/RendererLara.cpp

```diff
@@ -222,59 +229,57 @@ void Renderer::UpdateLaraAnimations(bool force)
+			auto rightFrame = leftAnim.GetKeyframeInterpData(rightAnimData).Keyframe0;
+
 			// Left arm
 			int upperArmMask = MESH_BITS(LM_LINARM);
 			mask = MESH_BITS(LM_LOUTARM) | MESH_BITS(LM_LHAND);
-			auto pistolFrameData = AnimFrameInterpData
-			{
-				&g_Level.Frames[Lara.LeftArm.FrameBase + Lara.LeftArm.FrameNumber - GetAnimData(Lara.LeftArm.AnimNumber).frameBase],
-				&g_Level.Frames[Lara.LeftArm.FrameBase + Lara.LeftArm.FrameNumber - GetAnimData(Lara.LeftArm.AnimNumber).frameBase],
-				0.0f
-			};
+			auto interpDataLeft = KeyframeInterpData(leftFrame, leftFrame, 0.0f);
 
-			UpdateAnimation(&rItem, playerObject, pistolFrameData, upperArmMask, true);
-			UpdateAnimation(&rItem, playerObject, pistolFrameData, mask);
+			UpdateAnimation(&rItem, playerObject, interpDataLeft, upperArmMask, true);
+			UpdateAnimation(&rItem, playerObject, interpDataLeft, mask);
 
 			// Right arm
 			upperArmMask = MESH_BITS(LM_RINARM);
 			mask = MESH_BITS(LM_ROUTARM) | MESH_BITS(LM_RHAND);
-			pistolFrameData = AnimFrameInterpData
-			{
-				&g_Level.Frames[Lara.RightArm.FrameBase + Lara.RightArm.FrameNumber - GetAnimData(Lara.RightArm.AnimNumber).frameBase],
-				&g_Level.Frames[Lara.RightArm.FrameBase + Lara.RightArm.FrameNumber - GetAnimData(Lara.RightArm.AnimNumber).frameBase],
-				0.0f
-			};
+			auto interpDataRight = KeyframeInterpData(rightFrame, rightFrame, 0.0f);
 			
-			UpdateAnimation(&rItem, playerObject, pistolFrameData, upperArmMask, true);
-			UpdateAnimation(&rItem, playerObject, pistolFrameData, mask);
+			UpdateAnimation(&rItem, playerObject, interpDataRight, upperArmMask, true);
+			UpdateAnimation(&rItem, playerObject, interpDataRight, mask);
 		}
 
 		break;
 
 		case LaraWeaponType::Flare:
 		case LaraWeaponType::Torch:
 			// Left arm
-			ItemInfo tempItem;
-			tempItem.Animation.AnimNumber = Lara.LeftArm.AnimNumber;
-			tempItem.Animation.FrameNumber = Lara.LeftArm.FrameNumber;
+			auto leftAnimData = GetNormalizedArmAnimFrame(Lara.LeftArm.AnimObjectID, Lara.LeftArm.FrameNumber);
+			const auto& leftAnim = GetAnimData(Lara.LeftArm.AnimObjectID, Lara.LeftArm.AnimNumber);
+			auto leftFrame = leftAnim.GetKeyframeInterpData(leftAnimData).Keyframe0;
 
 			mask = MESH_BITS(LM_LINARM) | MESH_BITS(LM_LOUTARM) | MESH_BITS(LM_LHAND);
 
-			// HACK: Mask head and torso when taking out a flare.
+			// HACK: Mask head and torso when taking out flare.
 			if (!Lara.Control.IsLow &&
-				tempItem.Animation.AnimNumber > (Objects[ID_FLARE_ANIM].animIndex + 1) &&
-				tempItem.Animation.AnimNumber < (Objects[ID_FLARE_ANIM].animIndex + 4))
+				Lara.LeftArm.AnimNumber > 1 &&
+				Lara.LeftArm.AnimNumber < 4)
 			{
 				mask |= MESH_BITS(LM_TORSO) | MESH_BITS(LM_HEAD);
 			}
 
-			auto frameData = GetFrameInterpData(tempItem);
-			UpdateAnimation(&rItem, playerObject, frameData, mask);
+			auto interpDataLeft = KeyframeInterpData(leftFrame, leftFrame, 0.0f);
```

Comment:
I am not sure why `tempItem` was used here, probably to bypass some Lara animation properties? Flare/torch wielding must be visually checked.